File: LeetCode/506-Relative Ranks/solution.py
#!/usr/local/bin/python3
# coding: UTF-8
# Author: David
# Email: youchen.du@gmail.com
# Created: 2017-02-13 14:15
# Last modified: 2017-02-13 14:39
# Filename: solution.py
# Description:
class Solution(object):
    def findRelativeRanks(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        ranks = list(range(1, len(nums)+1))
        ranks = [str(r) for r in ranks]
        ranks[:3] = ["Gold Medal", "Silver Medal", "Bronze Medal"]
        nis = [(num, idx) for idx, num in enumerate(nums)]
        nis.sort(key=lambda x: x[0], reverse=True)
        results = [None]*len(nums)
        for _idx, (num, idx) in enumerate(nis):
            results[idx] = ranks[_idx]
        return results

File: LeetCode/506-Relative Ranks/solution.cpp
// Author: David
// Email: youchen.du@gmail.com
// Created: 2017-02-13 14:56
// Last modified: 2017-02-13 14:56
// Filename: solution.cpp
// Description:
bool cmp(const pair<int, int>& a, const pair<int, int>& b)
{
    return a.first > b.first;
}
class Solution {
public:
    vector<string> findRelativeRanks(vector<int>& nums) {
        vector<pair<int, int>> nis;
        vector<string> ranks;
        vector<string> results;
        results.resize(nums.size());
        for(int i=0; i < nums.size(); i++)
        {
            if(i == 0)
                ranks.push_back("Gold Medal");
            else if (i == 1)
                ranks.push_back("Silver Medal");
            else if (i == 2)
                ranks.push_back("Bronze Medal");
            else
                ranks.push_back(to_string(i+1));
        }
        for(int i=0; i < nums.size(); i++)
            nis.push_back(make_pair(nums[i], i));
        sort(nis.begin(), nis.end(), cmp);
        
        for(int i=0; i < nums.size(); i++)
        {
            results[nis[i].second] = ranks[i];
        }
        return results;
    }
};
